File: seir-ten3/express-auth-api/controllers/booksController.js
const express = require('express');
const router = express.Router();

const Book = require('../models/Book');

// GET (index) /api/books/
router.get('/', (req, res) => {
  Book.find()
    .then((books) => res.json(books))
    .catch(console.error);
});

// GET (show) /api/books/5eb579b99b05e67b897e860b
router.get('/:id', (req, res) => {
  Book.findById(req.params.id)
    .then((book) => res.json(book))
    .catch(console.error);
});

// POST (create) /api/books/
router.post('/', (req, res) => {
  Book.create(req.body)
    .then((book) => res.status(201).json(book))
    .catch(console.error);
});

// PUT (update) /api/books/5eb579b99b05e67b897e860b
router.put('/:id', (req, res) => {
  Book.findOneAndUpdate({ _id: req.params.id }, req.body, {
    new: true,
  })
    .then((book) => res.json(book))
    .catch(console.error);
});

// DELETE (delete) /api/books/5eb579b99b05e67b897e860b
router.delete('/:id', (req, res) => {
  Book.findOneAndDelete({
    _id: req.params.id,
  })
    .then(() => res.sendStatus(204))
    .catch(console.error);
});

module.exports = router;

File: General Assembly WDI Boston/library-api-rails/app/controllers/books_controller.rb
# frozen_string_literal: true

# BooksController class
class BooksController < ApplicationController
  before_action :set_book, only: %i[show update destroy]

  # GET /books
  # GET /books.json
  def index
    @books = Book.all

    render json: @books
  end

  # GET /books/1
  # GET /books/1.json
  def show
    render json: @book
  end

  # POST /books
  # POST /books.json
  def create
    @book = Book.new(book_params)

    if @book.save
      render json: @book, status: :created
    else
      render json: @book.errors, status: :unprocessable_entity
    end
  end

  # PATCH/PUT /books/1
  # PATCH/PUT /books/1.json
  def update
    if @book.update(book_params)
      render json: @book
    else
      render json: @book.errors, status: :unprocessable_entity
    end
  end

  # DELETE /books/1
  # DELETE /books/1.json
  def destroy
    @book.destroy

    head :no_content
  end

  def set_book
    @book = Book.find(params[:id])
  end

  def book_params
    params.require(:book).permit(:title, :author)
  end

  # Return 418 for incoming POST with an id in the route, like POST /books/1
  # Intended to be an easter egg, and if devs are curious what / why and get here,
  # it will show how to make a custom route with custom status code response
  # More on 418: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418
  def teapot
    render json: { error: { message: 'You cannot POST with an ID, as the database provides the ID' } }, status: 418
  end

  private :set_book, :book_params
end
